### ISSUE TITLE
S3 exception while copying final outputs to s3 folder

### ISSUE THREAD

**Issue by doron-st** [broadinstitute/cromwell#4686]

**Comment by doron-st** [comment]
Here is the wdl file used for the test:

```
workflow three_task_sequence{
    call print_nach

    call print_nach_nachman {
        input:
            previous_task_out = print_nach.out
    }

    call print_nach_nachman_meuman{
        input:
            previous_task_out = print_nach_nachman.out
    }

    output{
        File out = print_nach_nachman_meuman.out
    }
}

task print_nach{

     String output_file = "task1.out"
     command{
         echo "nach" > ${output_file}
     }
     output{
         File out = output_file
     }
     runtime {
	    docker: "ubuntu:latest"
	    maxRetries: 3
     }
 }

task print_nach_nachman{
    File previous_task_out
    Array[String] previous_content = read_lines(previous_task_out)
    String output_file = "task2.out"
     command{
        echo ${sep=' ' previous_content} " nachman" > ${output_file}
     }
     output{
        File out = output_file
     }
     runtime {
        docker: "ubuntu:latest"
        maxRetries: 3
     }
     
 }

task print_nach_nachman_meuman{
    File previous_task_out
    Array[String] previous_content = read_lines(previous_task_out)
    String output_file = "three_task_sequence.out"
    command{
        echo ${sep=' ' previous_content} " meuman" > ${output_file}
    }
    output{
        File out = output_file
    }
    runtime {
        docker: "ubuntu:latest"
        maxRetries: 3
    }
  }
```

**Comment by geoffjentry** [comment]
Hi @doron-st Can you post your Cromwell config & how you set up your batch environment, e.g. via the cloud formation launch templates?

**Comment by hisplan** [comment]
I'm having the same issue. I've used the CloudFormation template as is. Didn't make any changes in the Cromwell config.

**Comment by wleepang** [comment]
This line in the stack trace indicates the root cause:

```
at software.amazon.awssdk.services.s3.S3Client.listBuckets(S3Client.java:2184)
```

The permissions for the Cromwell server currently only provide full access to the S3 bucket specified in the CloudFormation template.  Adding the `AmazonS3ReadOnlyAccess` managed policy to the server's instance profile fixes this, but I wonder if it can be more refined.  Specifically, is there a case where a Cromwell server would need to read from more than one bucket?

**Comment by ruchim** [comment]
Yes there are! Like if your inputs are shared with other users and they
come from varying sources and you're not allowed to make copies. Something
like that is very common.

**Comment by wleepang** [comment]
Probably worth distinguishing the difference between the server instance and individual job instances.  Each can get independent access permissions.  I can see the need for job instances to be able to read from various input sources, including those not owned by the account. Does the server, i.e. the Cromwell process, need to read data across accounts as well?  I was under the impression that it only needed to track status and logs for workflows and jobs submitted to it.

**Comment by allowishus** [comment]
Has anyone found a solution to this? I've experienced this issue since Cromwell was released for AWS and haven't found a good solution other than transferring all input files and references into the specific Cromwell S3 execution bucket.

**Comment by markjschreiber** [comment]
The role granted to the EC2 that runs the Cromwell server would need to grant access to the bucket that you have specified in the outputs.

**Comment by markjschreiber** [comment]
By default, the CloudFormation template will only give access to the bucket you specified at creation time as well as `gatk-test-data/*` and `broad-references/*`. To be able to access data in additional buckets you would need to grant `s3.*` to these resources through a policy that grants access to the bucket that you attach to `Cromwell-ServerStac-Ec2InstanceRole` (The exact name of the role depends on the name you gave the stack and some random characters cloud formation adds to prevent name collisions). In addition you need to add the same (or equivalent) policy to `GenomicsW-GenomicsEnvBatchInstance` role. This grants the batch worker EC2s access to the bucket.

The policy would look something like:
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": [
                "s3:*"
            ],
            "Resource": [
                "arn:aws:s3:::my-bucket-name",
                "arn:aws:s3:::my-bucket-name/*"
            ],
            "Effect": "Allow",
            "Sid": "S3BucketAllowAllObjectOps"
        }
    ]
}
```
